File: analytics/src/test/java/io/appmetrica/analytics/testutils/GlobalServiceLocatorRule.java
package io.appmetrica.analytics.testutils;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import androidx.annotation.NonNull;
import io.appmetrica.analytics.coreapi.internal.executors.IHandlerExecutor;
import io.appmetrica.analytics.coreapi.internal.executors.InterruptionSafeThread;
import io.appmetrica.analytics.coreapi.internal.identifiers.AppSetId;
import io.appmetrica.analytics.coreapi.internal.identifiers.AppSetIdScope;
import io.appmetrica.analytics.coreapi.internal.identifiers.PlatformIdentifiers;
import io.appmetrica.analytics.coreapi.internal.servicecomponents.applicationstate.ApplicationState;
import io.appmetrica.analytics.coreapi.internal.servicecomponents.applicationstate.ApplicationStateObserver;
import io.appmetrica.analytics.coreapi.internal.system.ActiveNetworkTypeProvider;
import io.appmetrica.analytics.coreapi.internal.system.PermissionExtractor;
import io.appmetrica.analytics.coreutils.internal.ReferenceHolder;
import io.appmetrica.analytics.coreutils.internal.services.FirstExecutionConditionServiceImpl;
import io.appmetrica.analytics.coreutils.internal.services.WaitForActivationDelayBarrier;
import io.appmetrica.analytics.impl.ApplicationStateProviderImpl;
import io.appmetrica.analytics.impl.BatteryInfoProvider;
import io.appmetrica.analytics.impl.ClidsInfoStorage;
import io.appmetrica.analytics.impl.DataSendingRestrictionControllerImpl;
import io.appmetrica.analytics.impl.ExtraMetaInfoRetriever;
import io.appmetrica.analytics.impl.GlobalServiceLocator;
import io.appmetrica.analytics.impl.LifecycleDependentComponentManager;
import io.appmetrica.analytics.impl.PreloadInfoStorage;
import io.appmetrica.analytics.impl.SdkEnvironmentHolder;
import io.appmetrica.analytics.impl.SelfDiagnosticReporterStorage;
import io.appmetrica.analytics.impl.StartupStateHolder;
import io.appmetrica.analytics.impl.component.ComponentId;
import io.appmetrica.analytics.impl.component.ServiceModuleReporterComponentLifecycleImpl;
import io.appmetrica.analytics.impl.crash.ndk.NativeCrashService;
import io.appmetrica.analytics.impl.db.VitalCommonDataProvider;
import io.appmetrica.analytics.impl.db.VitalComponentDataProvider;
import io.appmetrica.analytics.impl.db.VitalDataProviderStorage;
import io.appmetrica.analytics.impl.db.preferences.PreferencesServiceDbStorage;
import io.appmetrica.analytics.impl.id.AdvertisingIdGetter;
import io.appmetrica.analytics.impl.id.AppSetIdGetter;
import io.appmetrica.analytics.impl.location.LocationApi;
import io.appmetrica.analytics.impl.modules.ModuleEntryPointsRegister;
import io.appmetrica.analytics.impl.modules.ModuleEventHandlersHolder;
import io.appmetrica.analytics.impl.modules.service.ServiceModulesController;
import io.appmetrica.analytics.impl.network.http.SslSocketFactoryProviderImpl;
import io.appmetrica.analytics.impl.referrer.service.ReferrerHolder;
import io.appmetrica.analytics.impl.service.ServiceDataReporterHolder;
import io.appmetrica.analytics.impl.servicecomponents.ServiceLifecycleTimeTracker;
import io.appmetrica.analytics.impl.startup.CollectingFlags;
import io.appmetrica.analytics.impl.startup.StartupState;
import io.appmetrica.analytics.impl.startup.uuid.MultiProcessSafeUuidProvider;
import io.appmetrica.analytics.impl.telephony.TelephonyDataProvider;
import io.appmetrica.analytics.impl.utils.process.CurrentProcessDetector;
import io.appmetrica.analytics.impl.utils.executors.ExecutorWrapper;
import io.appmetrica.analytics.impl.utils.executors.InterruptionSafeHandlerThread;
import io.appmetrica.analytics.impl.utils.executors.ServiceExecutorProvider;
import io.appmetrica.analytics.networktasks.internal.NetworkCore;
import java.util.UUID;
import java.util.concurrent.Callable;
import java.util.concurrent.Executor;
import java.util.concurrent.Future;
import java.util.concurrent.FutureTask;
import java.util.concurrent.TimeUnit;
import org.junit.rules.ExternalResource;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class GlobalServiceLocatorRule extends ExternalResource {

    private static Answer sRunnableAnswer = new Answer() {
        @Override
        public Object answer(InvocationOnMock invocation) throws Throwable {
            ((Runnable) invocation.getArgument(0)).run();
            return null;
        }
    };
    private static StubbedExecutorWrapper sStubbedExecutor = new StubbedExecutorWrapper("Stubbed executor thread");
    private static Answer sStubbedThreadAnswer = new Answer() {
        @Override
        public Object answer(InvocationOnMock invocation) throws Throwable {
            return new InterruptionSafeThread(((Runnable) invocation.getArgument(0)));
        }
    };

    @Override
    public void before() {
        GlobalServiceLocator globalServiceLocator = mock(GlobalServiceLocator.class);
        ServiceExecutorProvider serviceExecutorProvider = createStubbedServiceThreadProvider();
        when(globalServiceLocator.getServiceExecutorProvider()).thenReturn(serviceExecutorProvider);
        when(globalServiceLocator.getServicePreferences()).thenReturn(mock(PreferencesServiceDbStorage.class));
        LifecycleDependentComponentManager lifecycleDependentComponentManager = mock(LifecycleDependentComponentManager.class);
        ApplicationStateProviderImpl applicationStateProvider = mock(ApplicationStateProviderImpl.class);
        when(applicationStateProvider.getCurrentState()).thenReturn(ApplicationState.UNKNOWN);
        when(applicationStateProvider.registerStickyObserver(any(ApplicationStateObserver.class))).thenReturn(ApplicationState.UNKNOWN);
        when(lifecycleDependentComponentManager.getApplicationStateProvider()).thenReturn(applicationStateProvider);
        when(globalServiceLocator.getApplicationStateProvider()).thenReturn(applicationStateProvider);
        when(globalServiceLocator.getLifecycleDependentComponentManager()).thenReturn(lifecycleDependentComponentManager);
        when(globalServiceLocator.getBatteryInfoProvider()).thenReturn(mock(BatteryInfoProvider.class));
        when(globalServiceLocator.getClidsStorage()).thenReturn(mock(ClidsInfoStorage.class));
        Context context = TestUtils.createMockedContext();
        when(globalServiceLocator.getContext()).thenReturn(context);
        when(globalServiceLocator.getPreloadInfoStorage()).thenReturn(mock(PreloadInfoStorage.class));
        when(globalServiceLocator.getReferrerHolder()).thenReturn(mock(ReferrerHolder.class));
        when(globalServiceLocator.getSelfDiagnosticReporterStorage()).thenReturn(mock(SelfDiagnosticReporterStorage.class));
        when(globalServiceLocator.getDataSendingRestrictionController()).thenReturn(mock(DataSendingRestrictionControllerImpl.class));
        when(globalServiceLocator.getSslSocketFactoryProvider()).thenReturn(mock(SslSocketFactoryProviderImpl.class));
        VitalDataProviderStorage vitalDataProviderStorage = mock(VitalDataProviderStorage.class);
        when(vitalDataProviderStorage.getCommonDataProvider()).thenReturn(mock(VitalCommonDataProvider.class));
        when(vitalDataProviderStorage.getCommonDataProviderForMigration()).thenReturn(mock(VitalCommonDataProvider.class));
        when(vitalDataProviderStorage.getComponentDataProvider(any(ComponentId.class))).thenReturn(mock(VitalComponentDataProvider.class));
        when(globalServiceLocator.getVitalDataProviderStorage()).thenReturn(vitalDataProviderStorage);
        when(globalServiceLocator.getModulesController()).thenReturn(mock(ServiceModulesController.class));
        when(globalServiceLocator.getGeneralPermissionExtractor()).thenReturn(mock(PermissionExtractor.class));
        StartupStateHolder startupStateHolder = mock(StartupStateHolder.class);
        when(startupStateHolder.getStartupState())
            .thenReturn(new StartupState.Builder(new CollectingFlags.CollectingFlagsBuilder().build()).build());
        when(globalServiceLocator.getStartupStateHolder()).thenReturn(startupStateHolder);
        LocationApi locationApi = mock(LocationApi.class);
        when(globalServiceLocator.getLocationClientApi()).thenReturn(locationApi);
        when(globalServiceLocator.getLocationServiceApi()).thenReturn(locationApi);
        when(globalServiceLocator.getServiceDataReporterHolder()).thenReturn(mock(ServiceDataReporterHolder.class));
        when(globalServiceLocator.getModuleEventHandlersHolder()).thenReturn(mock(ModuleEventHandlersHolder.class));
        TelephonyDataProvider telephonyDataProvider = mock(TelephonyDataProvider.class);
        when(globalServiceLocator.getTelephonyDataProvider()).thenReturn(telephonyDataProvider);
        when(globalServiceLocator.getModuleEntryPointsRegister()).thenReturn(mock(ModuleEntryPointsRegister.class));
        when(globalServiceLocator.getMultiProcessSafeUuidProvider()).thenReturn(mock(MultiProcessSafeUuidProvider.class));
        when(globalServiceLocator.getNativeCrashService()).thenReturn(mock(NativeCrashService.class));
        when(globalServiceLocator.getServicePreferences()).thenReturn(mock(PreferencesServiceDbStorage.class));
        when(globalServiceLocator.getNetworkCore()).thenReturn(mock(NetworkCore.class));
        when(globalServiceLocator.getSdkEnvironmentHolder()).thenReturn(mock(SdkEnvironmentHolder.class));
        AppSetIdGetter appSetIdGetter = mock(AppSetIdGetter.class);
        when(appSetIdGetter.getAppSetId()).thenReturn(new AppSetId(UUID.randomUUID().toString(), AppSetIdScope.APP));
        AdvertisingIdGetter advertisingIdGetter = mock(AdvertisingIdGetter.class);
        PlatformIdentifiers platformIdentifiers = mock(PlatformIdentifiers.class);
        when(platformIdentifiers.getAdvIdentifiersProvider()).thenReturn(advertisingIdGetter);
        when(platformIdentifiers.getAppSetIdProvider()).thenReturn(appSetIdGetter);
        when(globalServiceLocator.getAppSetIdGetter()).thenReturn(appSetIdGetter);
        when(globalServiceLocator.getAdvertisingIdGetter()).thenReturn(advertisingIdGetter);
        when(globalServiceLocator.getPlatformIdentifiers()).thenReturn(platformIdentifiers);
        when(globalServiceLocator.getActivationBarrier()).thenReturn(mock(WaitForActivationDelayBarrier.class));
        when(globalServiceLocator.getFirstExecutionConditionService())
            .thenReturn(mock(FirstExecutionConditionServiceImpl.class));
        when(globalServiceLocator.getExtraMetaInfoRetriever()).thenReturn(mock(ExtraMetaInfoRetriever.class));
        when(globalServiceLocator.getServiceLifecycleTimeTracker()).thenReturn(mock(ServiceLifecycleTimeTracker.class));
        when(globalServiceLocator.getCurrentProcessDetector()).thenReturn(mock(CurrentProcessDetector.class));
        when(globalServiceLocator.getReferenceHolder()).thenReturn(mock(ReferenceHolder.class));
        when(globalServiceLocator.getServiceModuleReporterComponentLifecycle())
            .thenReturn(mock(ServiceModuleReporterComponentLifecycleImpl.class));
        when(globalServiceLocator.getActiveNetworkTypeProvider()).thenReturn(mock(ActiveNetworkTypeProvider.class));
        GlobalServiceLocator.setInstance(globalServiceLocator);
    }

    @Override
    public void after() {
        GlobalServiceLocator.destroy();
    }

    private ServiceExecutorProvider createStubbedServiceThreadProvider() {
        ServiceExecutorProvider serviceExecutorProvider = mock(ServiceExecutorProvider.class);
        when(serviceExecutorProvider.getReportRunnableExecutor()).thenReturn(sStubbedExecutor);
        IHandlerExecutor moduleExecutor = createMockedExecutor();
        when(serviceExecutorProvider.getModuleExecutor()).thenReturn(moduleExecutor);
        when(serviceExecutorProvider.getNetworkTaskProcessorExecutor()).thenReturn(sStubbedExecutor);
        when(serviceExecutorProvider.getSupportIOExecutor()).thenReturn(sStubbedExecutor);
        when(serviceExecutorProvider.getDefaultExecutor()).thenReturn(sStubbedExecutor);
        when(serviceExecutorProvider.getUiExecutor()).thenReturn(sStubbedExecutor);
        when(serviceExecutorProvider.getMetricaCoreExecutor()).thenReturn(sStubbedExecutor);
        Executor blockingExecutorStub = mock(Executor.class);
        doAnswer(sRunnableAnswer).when(blockingExecutorStub).execute(any(Runnable.class));
        return serviceExecutorProvider;
    }

    private IHandlerExecutor createMockedExecutor() {
        IHandlerExecutor mock = mock(IHandlerExecutor.class);
        Answer<Void> blockingAnswer = new Answer<Void>() {
            @Override
            public Void answer(InvocationOnMock invocation) throws Throwable {
                ((Runnable) invocation.getArgument(0)).run();
                return null;
            }
        };
        doAnswer(blockingAnswer).when(mock).execute(any(Runnable.class));
        doAnswer(blockingAnswer).when(mock).executeDelayed(any(Runnable.class), anyLong());
        doAnswer(blockingAnswer).when(mock).executeDelayed(any(Runnable.class), anyLong(), any(TimeUnit.class));
        Handler handlerMock = mock(Handler.class);
        when(mock.getHandler()).thenReturn(handlerMock);
        Looper looperMock = mock(Looper.class);
        when(mock.getLooper()).thenReturn(looperMock);
        return mock;
    }

    static class StubbedExecutorWrapper extends ExecutorWrapper {

        public StubbedExecutorWrapper(@NonNull String threadName) {
            super(threadName);
        }

        public StubbedExecutorWrapper(@NonNull InterruptionSafeHandlerThread handlerThread,
                                      @NonNull Looper looper,
                                      @NonNull Handler handler) {
            super(handlerThread, looper, handler);
        }

        @NonNull
        @Override
        public Handler getHandler() {
            Handler handler = mock(Handler.class);
            doAnswer(sRunnableAnswer).when(handler).post(any(Runnable.class));
            doAnswer(sRunnableAnswer).when(handler).postDelayed(any(Runnable.class), anyLong());
            return handler;
        }

        @NonNull
        @Override
        public Looper getLooper() {
            return mock(Looper.class);
        }

        @Override
        public void execute(@NonNull Runnable runnable) {
            runnable.run();
        }

        @Override
        public <T> Future<T> submit(Callable<T> callable) {
            FutureTask<T> futureTask = new FutureTask<T>(callable);
            futureTask.run();
            return futureTask;
        }

        @Override
        public void executeDelayed(@NonNull Runnable runnable, long delay) {
            runnable.run();
        }

        @Override
        public void executeDelayed(@NonNull Runnable runnable, long delay, @NonNull TimeUnit timeUnit) {
            runnable.run();
        }

        @Override
        public void remove(@NonNull Runnable runnable) {
            //Do nothing
        }
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/testutils/GlobalServiceLocatorRule.kt
package io.appmetrica.analytics.testutils

import io.appmetrica.analytics.coreapi.internal.identifiers.AppSetId
import io.appmetrica.analytics.coreapi.internal.identifiers.AppSetIdScope
import io.appmetrica.analytics.coreapi.internal.identifiers.PlatformIdentifiers
import io.appmetrica.analytics.coreapi.internal.servicecomponents.applicationstate.ApplicationState
import io.appmetrica.analytics.coreapi.internal.servicecomponents.applicationstate.ApplicationStateObserver
import io.appmetrica.analytics.coreapi.internal.system.ActiveNetworkTypeProvider
import io.appmetrica.analytics.coreapi.internal.system.PermissionExtractor
import io.appmetrica.analytics.coreutils.internal.ReferenceHolder
import io.appmetrica.analytics.coreutils.internal.services.FirstExecutionConditionServiceImpl
import io.appmetrica.analytics.coreutils.internal.services.WaitForActivationDelayBarrier
import io.appmetrica.analytics.impl.ApplicationStateProviderImpl
import io.appmetrica.analytics.impl.BatteryInfoProvider
import io.appmetrica.analytics.impl.ClidsInfoStorage
import io.appmetrica.analytics.impl.DataSendingRestrictionControllerImpl
import io.appmetrica.analytics.impl.ExtraMetaInfoRetriever
import io.appmetrica.analytics.impl.GlobalServiceLocator
import io.appmetrica.analytics.impl.LifecycleDependentComponentManager
import io.appmetrica.analytics.impl.PreloadInfoStorage
import io.appmetrica.analytics.impl.SdkEnvironmentHolder
import io.appmetrica.analytics.impl.SelfDiagnosticReporterStorage
import io.appmetrica.analytics.impl.StartupStateHolder
import io.appmetrica.analytics.impl.component.ComponentId
import io.appmetrica.analytics.impl.component.ServiceModuleReporterComponentLifecycleImpl
import io.appmetrica.analytics.impl.crash.ndk.NativeCrashService
import io.appmetrica.analytics.impl.db.VitalCommonDataProvider
import io.appmetrica.analytics.impl.db.VitalComponentDataProvider
import io.appmetrica.analytics.impl.db.VitalDataProviderStorage
import io.appmetrica.analytics.impl.db.preferences.PreferencesServiceDbStorage
import io.appmetrica.analytics.impl.db.storage.ServiceStorageFactory
import io.appmetrica.analytics.impl.id.AdvertisingIdGetter
import io.appmetrica.analytics.impl.id.AppSetIdGetter
import io.appmetrica.analytics.impl.location.LocationApi
import io.appmetrica.analytics.impl.modules.ModuleEntryPointsRegister
import io.appmetrica.analytics.impl.modules.ModuleEventHandlersHolder
import io.appmetrica.analytics.impl.modules.service.ServiceModulesController
import io.appmetrica.analytics.impl.network.http.SslSocketFactoryProviderImpl
import io.appmetrica.analytics.impl.referrer.service.ReferrerHolder
import io.appmetrica.analytics.impl.service.ServiceDataReporterHolder
import io.appmetrica.analytics.impl.servicecomponents.ServiceLifecycleTimeTracker
import io.appmetrica.analytics.impl.startup.CollectingFlags
import io.appmetrica.analytics.impl.startup.StartupState
import io.appmetrica.analytics.impl.startup.uuid.MultiProcessSafeUuidProvider
import io.appmetrica.analytics.impl.telephony.TelephonyDataProvider
import io.appmetrica.analytics.impl.utils.executors.ServiceExecutorProvider
import io.appmetrica.analytics.impl.utils.process.CurrentProcessDetector
import io.appmetrica.analytics.networktasks.internal.NetworkCore
import org.junit.rules.ExternalResource
import org.mockito.kotlin.any
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import java.util.UUID

class GlobalServiceLocatorRule : ExternalResource() {

    val contextRule = ContextRule()
    val context by contextRule::context
    val reportRunnableExecutor = MockProvider.mockedBlockingExecutorMock()
    val moduleExecutor = MockProvider.mockedBlockingExecutorMock()
    val networkTaskProcessorExecutor = MockProvider.mockedBlockingExecutorMock()
    val supportIOExecutor = MockProvider.mockedBlockingExecutorMock()
    val defaultExecutor = MockProvider.mockedBlockingExecutorMock()
    val uiExecutor = MockProvider.mockedBlockingExecutorMock()
    val metricaCoreExecutor = MockProvider.mockedBlockingExecutorMock()

    val serviceExecutorProviderMock = mock<ServiceExecutorProvider> {
        on { reportRunnableExecutor } doReturn reportRunnableExecutor
        on { networkTaskProcessorExecutor } doReturn networkTaskProcessorExecutor
        on { moduleExecutor } doReturn moduleExecutor
        on { supportIOExecutor } doReturn supportIOExecutor
        on { defaultExecutor } doReturn defaultExecutor
        on { uiExecutor } doReturn uiExecutor
        on { metricaCoreExecutor } doReturn metricaCoreExecutor
    }

    override fun before() {
        contextRule.before()
        val globalServiceLocator = mock<GlobalServiceLocator>()
        whenever(globalServiceLocator.serviceExecutorProvider).thenReturn(serviceExecutorProviderMock)
        whenever(globalServiceLocator.getServicePreferences()).thenReturn(mock<PreferencesServiceDbStorage>())
        val lifecycleDependentComponentManager = mock<LifecycleDependentComponentManager>()
        val applicationStateProvider = mock<ApplicationStateProviderImpl>()
        whenever(applicationStateProvider.currentState).thenReturn(ApplicationState.UNKNOWN)
        whenever(applicationStateProvider.registerStickyObserver(any<ApplicationStateObserver>())).thenReturn(
            ApplicationState.UNKNOWN
        )
        whenever(lifecycleDependentComponentManager.applicationStateProvider).thenReturn(applicationStateProvider)
        whenever(globalServiceLocator.applicationStateProvider).thenReturn(applicationStateProvider)
        whenever(globalServiceLocator.getLifecycleDependentComponentManager())
            .thenReturn(lifecycleDependentComponentManager)
        whenever(globalServiceLocator.getBatteryInfoProvider()).thenReturn(mock<BatteryInfoProvider>())
        whenever(globalServiceLocator.getClidsStorage()).thenReturn(mock<ClidsInfoStorage>())
        whenever(globalServiceLocator.context).thenReturn(contextRule.context)
        whenever(globalServiceLocator.getPreloadInfoStorage()).thenReturn(mock<PreloadInfoStorage>())
        whenever(globalServiceLocator.getReferrerHolder()).thenReturn(mock<ReferrerHolder>())
        whenever(globalServiceLocator.selfDiagnosticReporterStorage).thenReturn(mock<SelfDiagnosticReporterStorage>())
        whenever(globalServiceLocator.getDataSendingRestrictionController()).thenReturn(
            mock<DataSendingRestrictionControllerImpl>()
        )
        whenever(globalServiceLocator.getSslSocketFactoryProvider()).thenReturn(mock<SslSocketFactoryProviderImpl>())
        val vitalDataProviderStorage = mock<VitalDataProviderStorage>()
        whenever(vitalDataProviderStorage.commonDataProvider).thenReturn(mock<VitalCommonDataProvider>())
        whenever(vitalDataProviderStorage.commonDataProviderForMigration).thenReturn(mock<VitalCommonDataProvider>())
        whenever(vitalDataProviderStorage.getComponentDataProvider(any<ComponentId>())).thenReturn(
            mock<VitalComponentDataProvider>()
        )
        whenever(globalServiceLocator.getVitalDataProviderStorage()).thenReturn(vitalDataProviderStorage)
        whenever(globalServiceLocator.getModulesController()).thenReturn(mock<ServiceModulesController>())
        whenever(globalServiceLocator.getGeneralPermissionExtractor()).thenReturn(mock<PermissionExtractor>())
        val startupStateHolder = mock<StartupStateHolder>()
        whenever(startupStateHolder.getStartupState())
            .thenReturn(StartupState.Builder(CollectingFlags.CollectingFlagsBuilder().build()).build())
        whenever(globalServiceLocator.startupStateHolder).thenReturn(startupStateHolder)
        val locationApi = mock<LocationApi>()
        whenever(globalServiceLocator.locationClientApi).thenReturn(locationApi)
        whenever(globalServiceLocator.locationServiceApi).thenReturn(locationApi)
        whenever(globalServiceLocator.serviceDataReporterHolder).thenReturn(mock<ServiceDataReporterHolder>())
        whenever(globalServiceLocator.moduleEventHandlersHolder).thenReturn(mock<ModuleEventHandlersHolder>())
        val telephonyDataProvider = mock<TelephonyDataProvider>()
        whenever(globalServiceLocator.getTelephonyDataProvider()).thenReturn(telephonyDataProvider)
        whenever(globalServiceLocator.moduleEntryPointsRegister).thenReturn(mock<ModuleEntryPointsRegister>())
        whenever(globalServiceLocator.getMultiProcessSafeUuidProvider())
            .thenReturn(mock<MultiProcessSafeUuidProvider>())
        whenever(globalServiceLocator.nativeCrashService).thenReturn(mock<NativeCrashService>())
        whenever(globalServiceLocator.getServicePreferences()).thenReturn(mock<PreferencesServiceDbStorage>())
        whenever(globalServiceLocator.networkCore).thenReturn(mock<NetworkCore>())
        whenever(globalServiceLocator.getSdkEnvironmentHolder()).thenReturn(mock<SdkEnvironmentHolder>())
        val appSetIdGetter = mock<AppSetIdGetter>()
        whenever(appSetIdGetter.getAppSetId()).thenReturn(AppSetId(UUID.randomUUID().toString(), AppSetIdScope.APP))
        val advertisingIdGetter = mock<AdvertisingIdGetter>()
        val platformIdentifiers = mock<PlatformIdentifiers>()
        whenever(platformIdentifiers.advIdentifiersProvider).thenReturn(advertisingIdGetter)
        whenever(platformIdentifiers.appSetIdProvider).thenReturn(appSetIdGetter)
        whenever(globalServiceLocator.getAppSetIdGetter()).thenReturn(appSetIdGetter)
        whenever(globalServiceLocator.getAdvertisingIdGetter()).thenReturn(advertisingIdGetter)
        whenever(globalServiceLocator.getPlatformIdentifiers()).thenReturn(platformIdentifiers)
        whenever(globalServiceLocator.activationBarrier).thenReturn(mock<WaitForActivationDelayBarrier>())
        whenever(globalServiceLocator.firstExecutionConditionService)
            .thenReturn(mock<FirstExecutionConditionServiceImpl>())
        whenever(globalServiceLocator.getExtraMetaInfoRetriever()).thenReturn(mock<ExtraMetaInfoRetriever>())
        whenever(globalServiceLocator.serviceLifecycleTimeTracker).thenReturn(mock<ServiceLifecycleTimeTracker>())
        whenever(globalServiceLocator.getCurrentProcessDetector()).thenReturn(mock<CurrentProcessDetector>())
        whenever(globalServiceLocator.referenceHolder).thenReturn(mock<ReferenceHolder>())
        whenever(globalServiceLocator.serviceModuleReporterComponentLifecycle)
            .thenReturn(mock<ServiceModuleReporterComponentLifecycleImpl>())
        whenever(globalServiceLocator.activeNetworkTypeProvider).thenReturn(mock<ActiveNetworkTypeProvider>())
        whenever(globalServiceLocator.getStorageFactory()).thenReturn(mock<ServiceStorageFactory>())
        GlobalServiceLocator.setInstance(globalServiceLocator)
    }

    override fun after() {
        GlobalServiceLocator.destroy()
        contextRule.after()
    }
}
